core/src/Kokkos_Half.hpp

```diff
@@ -0,0 +1,111 @@
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+// PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL NTESS OR THE
+// CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+// EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+// PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+// PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
+// LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+// NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+// SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+// Questions? Contact Christian R. Trott (crtrott@sandia.gov)
+//
+// ************************************************************************
+//@HEADER
+*/
+
+#ifndef KOKKOS_HALF_HPP_
+#define KOKKOS_HALF_HPP_
+
+#include <type_traits>
+#include <cstdint>
+#include <Kokkos_Macros.hpp>
+
+// Include special backend specific versions here
+#include <Cuda/Kokkos_Cuda_Half.hpp>
+
+// Potentially include special compiler specific versions here
+// e.g. for Intel
+
+// If none of the above actually did anything and defined a half precision type
+// define a fallback implementation here using float
+#ifndef KOKKOS_IMPL_HALF_TYPE_DEFINED
+#define KOKKOS_IMPL_HALF_TYPE_DEFINED
+namespace Kokkos {
+namespace Experimental {
+
+using half_t                       = float;
+constexpr const bool half_is_float = true;
+
+// cast_to_half
+KOKKOS_INLINE_FUNCTION
+half_t cast_to_half(float val) { return half_t(val); }
+KOKKOS_INLINE_FUNCTION
+half_t cast_to_half(double val) { return half_t(val); }
+KOKKOS_INLINE_FUNCTION
+half_t cast_to_half(short val) { return half_t(val); }
+KOKKOS_INLINE_FUNCTION
+half_t cast_to_half(unsigned short val) { return half_t(val); }
+KOKKOS_INLINE_FUNCTION
+half_t cast_to_half(int val) { return half_t(val); }
+KOKKOS_INLINE_FUNCTION
+half_t cast_to_half(unsigned int val) { return half_t(val); }
+KOKKOS_INLINE_FUNCTION
+half_t cast_to_half(long val) { return half_t(val); }
+KOKKOS_INLINE_FUNCTION
+half_t cast_to_half(unsigned long val) { return half_t(val); }
+KOKKOS_INLINE_FUNCTION
+half_t cast_to_half(long long val) { return half_t(val); }
+KOKKOS_INLINE_FUNCTION
+half_t cast_to_half(unsigned long long val) { return half_t(val); }
+
+// cast_from_half
+// Using an explicit list here too, since the other ones are explicit and for
+// example don't include char
+template <class T>
+KOKKOS_INLINE_FUNCTION typename std::enable_if<
+    std::is_same<T, float>::value || std::is_same<T, double>::value ||
+        std::is_same<T, short>::value ||
+        std::is_same<T, unsigned short>::value || std::is_same<T, int>::value ||
+        std::is_same<T, unsigned int>::value || std::is_same<T, long>::value ||
+        std::is_same<T, unsigned long>::value ||
+        std::is_same<T, long long>::value ||
+        std::is_same<T, unsigned long long>::value,
```

Comment:
consider doing
```C++
namespace Impl {
template <class T>
struct is_in_my_list : std::integral_constant<bool, std::is_same<T, int>::value || ...> {};
}
template <class T>
KOKKOS_INLINE_FUNCTION std::enable_if_t<Impl::is_in_my_list<T>::value, T>
cast_from_half(half_t val) { return T(val); }

template <class T>
KOKKOS_INLINE_FUNCTION std::enable_if_t<Impl::is_in_my_list<T>::value, half_t>
cast_to_half(half_t val) { return half_t(val); }
```

Reply:
I thought about it, actually but I liked that this listing was closer to the visuals in some way to the other variants.

Reply:
Christian made the case that `cast_to_half` is not templated in the CUDA backend implementation and that his intent was to mirror it.